dotnet/src/SemanticKernel.Abstractions/AI/PromptExecutionSettings.cs

```diff
@@ -39,5 +52,43 @@ public class PromptExecutionSettings
     /// Avoid using this property if possible. Instead, use one of the classes that extends <see cref="PromptExecutionSettings"/>.
     /// </remarks>
     [JsonExtensionData]
-    public Dictionary<string, object>? ExtensionData { get; set; }
+    public IDictionary<string, object>? ExtensionData
+    {
+        get => this._extensionData;
+
+        set
+        {
+            if (this._isFrozen)
+            {
+                throw new InvalidOperationException("PromptExecutionSettings are frozen and ExtensionData cannot be modified.");
+            }
+            this._extensionData = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets a value that indicates whether the <see cref="PromptExecutionSettings"/> are currently modifiable.
+    /// </summary>
+    public bool IsFrozen
+    {
+        get => this._isFrozen;
+    }
+
+    /// <summary>
+    /// Makes the current <see cref="PromptExecutionSettings"/> unmodifiable and sets its IsFrozen property to true.
+    /// </summary>
+    public PromptExecutionSettings Freeze()
```

Comment:
I think this should be virtual so that derived implementations can react to it.

For example, this base implementation should probably do:
```C#
if (_extensionData is not null)
{
    _extensionData = new ReadOnlyDictionary<string, object?>(_extensionData);
}
```
or something like that in the Freeze call.